pkg/aggregateeventrecorder/aggregateeventrecorder.go

```diff
@@ -0,0 +1,215 @@
+}
+
+type batchedEvents struct {
+	Events []AggregateEvent `json:"events"`
+}
+
+type aggregateEventRecorder struct {
+	ctx                   context.Context
+	instanceID            string                    // The ID of the instance generating the event
+	endpointURL           string                    // The URL to POST the events to
+	endpointAuthorization string                    // The key to use in an Authorization header
+	disableIndexerEvents  bool                      // Whether or not to disable indexer events. Defaults to false.
+	ingestChan            chan types.RetrievalEvent // A channel for incoming events
+	postChan              chan []AggregateEvent     // A channel for posting events
+	eventTempMap          map[string]tempData       // Holds data about multiple events for aggregation
+}
+
+func NewAggregateEventRecorder(ctx context.Context, instanceID string, endpointURL string, endpointAuthorization string, disableIndexerEvents bool) *aggregateEventRecorder {
+	recorder := &aggregateEventRecorder{
+		ctx:                   ctx,
+		instanceID:            instanceID,
+		endpointURL:           endpointURL,
+		endpointAuthorization: endpointAuthorization,
+		disableIndexerEvents:  disableIndexerEvents,
+		ingestChan:            make(chan types.RetrievalEvent),
+		postChan:              make(chan []AggregateEvent),
+		eventTempMap:          make(map[string]tempData),
+	}
+
+	go recorder.ingestEvents()
+	for i := 0; i < parallelPosters; i++ {
+		go recorder.postEvents()
+	}
+
+	return recorder
+}
+
+// RetrievalEventSubsciber returns a RetrievalEventSubscriber that POSTs
+// a batch of aggregated retrieval events to an event recorder API endpoint
+func (a *aggregateEventRecorder) RetrievalEventSubscriber() types.RetrievalEventSubscriber {
+	return func(event types.RetrievalEvent) {
+		if a.disableIndexerEvents && event.Phase() == types.IndexerPhase {
+			return
+		}
+
+		// Process the incoming event
+		select {
+		case <-a.ctx.Done():
+		case a.ingestChan <- event:
+		}
+	}
+}
+
+// ingestEvents receives and stores event data from ingestChan
+// to generate an aggregated event upon receiving a finished event.
+func (a *aggregateEventRecorder) ingestEvents() {
+	var batchedData []AggregateEvent
+	var emptyGaurdChan chan []AggregateEvent = nil
+
+	for {
+		select {
+		case <-a.ctx.Done():
+		// read incoming data
+		case event := <-a.ingestChan:
+			id := event.RetrievalId().String()
+
+			switch event.Code() {
+			case types.StartedCode:
+				// Record when first phase starts to help calculate time to first byte later
+				if event.Phase() == types.FetchPhase {
+					a.eventTempMap[id] = tempData{
+						startTime:     event.Time(),
+						firstByteTime: time.Time{},
+						spId:          "",
+						success:       false,
+						bandwidth:     0,
+						ttfb:          0,
+					}
+				}
+
+			case types.FirstByteCode:
+				// Calculate time to first byte
```

Comment:
need to check if first byte is already written. multiple retrievals could get to first byte

Reply:
After some discussion, we're going push an always 0 ttfb and bandwidth for now due to work that requires a similar solution in https://github.com/filecoin-project/lassie/issues/152.